Basic Crash Reporter

magmaus3 opened the pull request:
This PR adds an crash reporter + saving logs to files.

 - Log uploads are using transfer.sh for storage.
 - When the app crashes, an popup (currently using tkinter) appears, asking if log should be sent.
   Then it prints out the log url in command line.
 - Logs on the user's machine are automatically archived on launch.

Commit:
Logging to file and crash reporting

Commit:
Send logs on crash
	- Ignores KeyboardInterrupt and EOFError
	- Currently works automatically, this will be changed later

Commit:
Popup requesting sending logs permission + logging system information

Commit:
Archive previous log files

Commit:
Open the log link in the new tab

bitcraft commented on run_tuxemon.py:
```diff
@@ -52,4 +55,16 @@
+        
+        logging.error(ex, exc_info=True)
+        if ex != KeyboardInterrupt and ex != EOFError:
+            #if input("Send logs? (Y/N)\n> ").upper() == "Y":
```
is this needed?

magmaus3 replied:
The logging.error part is for printing the traceback, so it isn't "eaten" by the except function.

The commented if is not needed, I'll remove it later.

bitcraft commented on tuxemon/config.py:
```diff
@@ -119,6 +119,12 @@ def __init__(self, config_path: Optional[str] = None) -> None:
         self.debug_level = cfg.get("logging", "debug_level")
 
+        
+        self.log_host_url = cfg.get("crash_reporting", "log_host_url")  # URL to transfer.sh compatible
```
please put the comments above the line, it looks much cleaner

bitcraft commented on tuxemon/log.py:
```diff
@@ -32,8 +32,34 @@
+# For gathering system information
+import platform
 
+def archive_log():
```
this is fine now, but just a FYI, python has a log rotation, compression, etc.  if this gets more complex, we should probably use what is in the stdlib
https://stackoverflow.com/questions/8467978/python-want-logging-with-log-rotation-and-compression

bitcraft commented on tuxemon/log.py:
```diff
@@ -72,8 +98,78 @@ def configure():
+    # Get file
+    file = open(f"{USER_LOG_DIR}/latest.log")
+    send_files = {"tuxemon_log.txt": file}
+    r = requests.post(config.log_host_url, files=send_files, headers={"Max-Days": config.log_storage_max_days, "User-Agent": "curl/7.80.0"})
```
im guessing that the header here is specific to transfer.sh?  to prevent leaking logs without proper expiration, lets either make a generic class that handles sending logs to a provider, or simply raise an error if its not going to transfer.sh, and put a TODO here about allowing other providers

bitcraft commented on tuxemon/config.py:
```diff
@@ -242,6 +248,16 @@ def get_defaults() -> Mapping[str, Any]:
+                "crash_reporting",
+                OrderedDict(
+                    (
+                        ("log_host_url", "https://transfer.sh/"),
```
could we add an option to enable the popup?  during development, its going to crash often, and the popup will get really annoying.

magmaus3 replied:
I was thinking about adding an option to the config with `None` as the default value (and creating the popup), and bool values for automatically sending/not sending logs.

bitcraft reviewed (commented):
looks great!  looking forward to seeing the logs being generated.  few comments

Commit:
Smaller changes

Commit:
Add an option to show the crash popup.
If None, always show the popup

magmaus3 reviewed (commented):
‮

Commit:
Fix one mistake

Commit:
LogStorageProvider generic class

Commit:
Change the log_host_url to log_host

Commit:
Attempting to fix a merge conflict
Merge branch 'crash-reporter' of https://github.com/vXtreniusX/Tuxemon into crash-reporter

Commit:
Merge branch 'Tuxemon:development' into crash-reporter

Commit:
Docstrings

magmaus3 commented:
The generic class is done

Commit:
Fix a bug

Commit:
Make the changes usable

Commit:
Changes

Commit:
Some fixes

Commit:
Use the log_host_max_storage_days variable from config

Commit:
Add a config option to disable opening links automatically

Commit:
Remove debug prints

bitcraft commented on tuxemon/log.py:
```diff
@@ -32,10 +32,94 @@
+        self.log_storage_max_days = storage_time
+
+
+
```
PEP8: too many lines between method blocks.  should be exactly one line

bitcraft commented on tuxemon/log.py:
```diff
@@ -46,34 +130,123 @@ def configure():
+    finally:
+        pass
+
+    # Get file
```
please avoid including commented code, especially if there is no note about what it is.

bitcraft commented on tuxemon/log.py:
```diff
@@ -46,34 +130,123 @@ def configure():
+
+    # TODO: Add other providers support
+    try:
+        provider = eval(f"{config.log_host}(storage_time={config.log_storage_max_days})")
```
what is eval being used here for?  can we avoid using it?

magmaus3 replied:
I used it for "spawning" a class object.

bitcraft commented on tuxemon/log.py:
```diff
@@ -46,34 +130,123 @@ def configure():
+    if prepare.CONFIG.popup:
+        return True
+        
+    elif not prepare.CONFIG.popup:
```
why not if/else?

bitcraft commented on tuxemon/log.py:
```diff
@@ -46,34 +130,123 @@ def configure():
+    except requests.exceptions.ConnectionError:
+        logging.error("Connection to the provider failed.", exc_info=True)
+    
+
```
PEP8: should be 2 lines between functions/classes defined in the global scope (not indented)

bitcraft commented on tuxemon/log.py:
```diff
@@ -32,10 +32,94 @@
+        """
+        files = {"tuxemon_log.txt": file_obj.read()}
+        r = requests.post(
+                            self.url,
```
excessive indentation

bitcraft commented on tuxemon/log.py:
```diff
@@ -46,34 +130,123 @@ def configure():
+        provider = eval(f"{config.log_host}(storage_time={config.log_storage_max_days})")
+        response = provider.send_log()
+        print(f"Report URL: {response[0]}")
+        if config.auto_open_link.upper()[0] == "T":  # Using the standard "if config" didn't work for me
```
we're trying to keep the lines short.  please avoid in line comments, especially when they create very long lines.  this one is 104 chars.

bitcraft commented on tuxemon/config.py:
```diff
@@ -119,6 +119,14 @@ def __init__(self, config_path: Optional[str] = None) -> None:
         self.debug_level = cfg.get("logging", "debug_level")
 
+        # [crash_reporting]        
+        self.log_host = cfg.get("crash_reporting", "log_host")  # The log storage type.
```
please move the comments to the line above the code

bitcraft reviewed (approved):
logic looks great, thank you for the changes.  i added more comments, mostly about formatting.  we are going to try to tighten up the formatting requirements.  
if you would like to have it done automatically, you can use 
```shell
black -t py38 -l 79
```
which is what our "official" formatter does.

bitcraft reviewed (changes requested):
see comments

Commit:
Auto Formatting

Commit:
Fixing merge conflict

Commit:
Comment changes

Commit:
Revert "Fixing merge conflict"

This reverts commit 66c399b9d1abc39bfd00fd48b05aff36ba4c2573, reversing
changes made to 0387aee8517c5c0cd2f62e19551e6a8b2e7b37ef.

Commit:
Merge branch 'development' into crash-reporter

Trying to fix a merge conflict

Commit:
Fixing merge request (I hope it will work correctly this time)

Commit:
Revert "Fixing merge conflict"

This reverts commit 66c399b9d1abc39bfd00fd48b05aff36ba4c2573, reversing
changes made to 0387aee8517c5c0cd2f62e19551e6a8b2e7b37ef.

Commit:
Revert "Revert "Fixing merge conflict""

This reverts commit d6e2feace90c37b62aa9fd24c89d6b77a47c9711.

Commit:
Revert "Merge branch 'development' into crash-reporter"

This reverts commit 36475408056edf3f52d3596e6883ca80c5d0579b, reversing
changes made to d6e2feace90c37b62aa9fd24c89d6b77a47c9711.

Commit:
Get provider class using a dict instead of eval()

Commit:
Make consent popup work properly

Commit:
indexation

magmaus3 commented:
Ready for review, note that I've encountered some merge conflicts, that I wasn't able to fix before. Not sure if I resolved them correctly now.

Sorry for the delay.

Commit:
Merge branch 'development' into crash-reporter
